File: combinatorics/resource/LCHP/problem_LTDT_code/bai_9/bai9.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <set>
#include <map>
#include <climits>
#include <iomanip>
#include <algorithm>

using namespace std;

class MultigraphBFS {
private:
    int numVertices;
    // Sử dụng multiset để lưu nhiều cạnh giữa hai đỉnh
    vector<multiset<int>> adjList;
    vector<bool> visited;
    vector<int> distance;
    vector<int> parent;
    vector<int> bfsOrder;
    
    // Đếm số cạnh giữa hai đỉnh
    map<pair<int,int>, int> edgeCount;
    
public:
    MultigraphBFS(int n) : numVertices(n) {
        adjList.resize(n);
        resetArrays();
    }
    
    void resetArrays() {
        visited.assign(numVertices, false);
        distance.assign(numVertices, INT_MAX);
        parent.assign(numVertices, -1);
        bfsOrder.clear();
    }
    
    // Thêm cạnh vào multigraph
    void addEdge(int u, int v) {
        adjList[u].insert(v);
        adjList[v].insert(u);
        
        // Đếm số cạnh
        pair<int,int> edge = {min(u,v), max(u,v)};
        edgeCount[edge]++;
        
        cout << "Them canh (" << u << "," << v << ") - So canh giua " 
             << u << " va " << v << ": " << edgeCount[edge] << endl;
    }
    
    // BFS cho multigraph
    void bfs(int start) {
        resetArrays();
        
        queue<int> q;
        visited[start] = true;
        distance[start] = 0;
        q.push(start);
        
        cout << "BFS tren Multigraph tu dinh " << start << ":" << endl;
        cout << "Thu tu duyet: ";
        
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            
            bfsOrder.push_back(u);
            cout << u << " ";
            
            // Sử dụng set để tránh duyệt trùng đỉnh kề
            set<int> uniqueNeighbors(adjList[u].begin(), adjList[u].end());
            
            for (int v : uniqueNeighbors) {
                if (!visited[v]) {
                    visited[v] = true;
                    distance[v] = distance[u] + 1;
                    parent[v] = u;
                    q.push(v);
                }
            }
        }
        cout << endl << endl;
    }
    
    // BFS cho toàn bộ multigraph (xử lý thành phần liên thông)
    void bfsComplete() {
        resetArrays();
        
        cout << "BFS toan bo multigraph:" << endl;
        int componentCount = 0;
        
        for (int i = 0; i < numVertices; i++) {
            if (!visited[i]) {
                componentCount++;
                cout << "Thanh phan lien thong " << componentCount << ": ";
                bfsComponent(i);
                cout << endl;
            }
        }
        
        cout << "Tong so thanh phan lien thong: " << componentCount << endl << endl;
    }
    
    // BFS cho một thành phần liên thông
    void bfsComponent(int start) {
        queue<int> q;
        visited[start] = true;
        distance[start] = 0;
        q.push(start);
        
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            cout << u << " ";
            
            set<int> uniqueNeighbors(adjList[u].begin(), adjList[u].end());
            for (int v : uniqueNeighbors) {
                if (!visited[v]) {
                    visited[v] = true;
                    distance[v] = distance[u] + 1;
                    parent[v] = u;
                    q.push(v);
                }
            }
        }
    }
    
    // Tìm đường đi ngắn nhất từ start đến target
    vector<int> getShortestPath(int start, int target) {
        bfs(start);
        
        vector<int> path;
        if (distance[target] == INT_MAX) {
            return path; // Không có đường đi
        }
        
        // Tái tạo đường đi từ mảng parent
        int current = target;
        while (current != -1) {
            path.push_back(current);
            current = parent[current];
        }
        
        reverse(path.begin(), path.end());
        return path;
    }
    
    // In thông tin về các cạnh bội
    void printMultipleEdges() {
        cout << "Cac canh boi trong multigraph:" << endl;
        bool hasMultiple = false;
        
        for (auto& edge : edgeCount) {
            if (edge.second > 1) {
                cout << "Canh (" << edge.first.first << "," 
                     << edge.first.second << "): " << edge.second << " canh" << endl;
                hasMultiple = true;
            }
        }
        
        if (!hasMultiple) {
            cout << "Khong co canh boi nao." << endl;
        }
        cout << endl;
    }
    
    // In thông tin khoảng cách
    void printDistances(int start) {
        cout << "Khoang cach tu dinh " << start << ":" << endl;
        cout << setw(6) << "Dinh" << setw(12) << "Khoang cach" << setw(10) << "Cha" << endl;
        cout << string(28, '-') << endl;
        
        for (int i = 0; i < numVertices; i++) {
            cout << setw(6) << i;
            if (distance[i] == INT_MAX) {
                cout << setw(12) << "INF";
            } else {
                cout << setw(12) << distance[i];
            }
            cout << setw(10) << parent[i] << endl;
        }
        cout << endl;
    }
    
    // In cây BFS
    void printBFSTree(int start) {
        cout << "Cay BFS tu dinh " << start << ":" << endl;
        
        bool hasTree = false;
        for (int i = 0; i < numVertices; i++) {
            if (parent[i] != -1) {
                cout << parent[i] << " -> " << i << endl;
                hasTree = true;
            }
        }
        
        if (!hasTree) {
            cout << "Khong co cay BFS (chi co dinh goc)." << endl;
        }
        cout << endl;
    }
    
    // Kiểm tra liên thông
    bool isConnected() {
        bfs(0);
        
        for (int i = 0; i < numVertices; i++) {
            if (!visited[i]) {
                return false;
            }
        }
        return true;
    }
    
    // Lấy số lượng cạnh giữa hai đỉnh
    int getEdgeMultiplicity(int u, int v) {
        pair<int,int> edge = {min(u,v), max(u,v)};
        return edgeCount.count(edge) ? edgeCount[edge] : 0;
    }
    
    // In đồ thị
    void printGraph() {
        cout << "Bieu dien Multigraph (danh sach ke):" << endl;
        for (int i = 0; i < numVertices; i++) {
            cout << i << ": ";
            for (int neighbor : adjList[i]) {
                cout << neighbor << " ";
            }
            cout << endl;
        }
        cout << endl;
    }
    
    // Thống kê đồ thị
    void printGraphStats() {
        cout << "Thong ke Multigraph:" << endl;
        cout << "So dinh: " << numVertices << endl;
        
        int totalEdges = 0;
        int uniqueEdges = edgeCount.size();
        
        for (auto& edge : edgeCount) {
            totalEdges += edge.second;
        }
        
        cout << "Tong so canh (ke ca canh boi): " << totalEdges << endl;
        cout << "So cap dinh co canh: " << uniqueEdges << endl;
        cout << "So canh boi: " << (totalEdges - uniqueEdges) << endl;
        cout << endl;
    }
};

// Các hàm tiện ích cho Multigraph
class MultigraphUtils {
public:
    // Tạo multigraph mẫu 1
    static MultigraphBFS createSampleMultigraph1() {
        MultigraphBFS graph(5);
        
        // Thêm cạnh bình thường
        graph.addEdge(0, 1);
        graph.addEdge(1, 2);
        graph.addEdge(2, 3);
        
        // Thêm cạnh bội
        graph.addEdge(0, 1); // Cạnh thứ 2 giữa 0 và 1
        graph.addEdge(1, 2); // Cạnh thứ 2 giữa 1 và 2
        graph.addEdge(3, 4);
        
        return graph;
    }
    
    // Tạo multigraph mẫu 2 - không liên thông
    static MultigraphBFS createSampleMultigraph2() {
        MultigraphBFS graph(6);
        
        // Thành phần 1
        graph.addEdge(0, 1);
        graph.addEdge(0, 1); // Cạnh bội
        graph.addEdge(1, 2);
        
        // Thành phần 2
        graph.addEdge(3, 4);
        graph.addEdge(3, 4); // Cạnh bội
        graph.addEdge(3, 4); // Cạnh bội thứ 2
        
        // Đỉnh cô lập
        // Đỉnh 5 không có cạnh nào
        
        return graph;
    }
};

// Hàm demo đầy đủ cho Multigraph
void demonstrateMultigraphBFS() {
    cout << "=== DEMO THUAT TOAN BFS TREN MULTIGRAPH ===" << endl << endl;
    
    // Demo 1: Multigraph liên thông với cạnh bội
    cout << "1. MULTIGRAPH LIEN THONG VOI CANH BOI:" << endl;
    auto graph1 = MultigraphUtils::createSampleMultigraph1();
    
    graph1.printGraph();
    graph1.printGraphStats();
    graph1.printMultipleEdges();
    
    // BFS từ đỉnh 0
    graph1.bfs(0);
    graph1.printDistances(0);
    graph1.printBFSTree(0);
    
    // Tìm đường đi ngắn nhất
    cout << "Duong di ngan nhat tu 0 den 4:" << endl;
    auto path = graph1.getShortestPath(0, 4);
    if (!path.empty()) {
        for (int i = 0; i < path.size(); i++) {
            cout << path[i];
            if (i < path.size() - 1) cout << " -> ";
        }
        cout << " (Do dai: " << path.size() - 1 << ")" << endl;
    } else {
        cout << "Khong co duong di!" << endl;
    }
    cout << endl;
    
    // Kiểm tra liên thông
    cout << "Do thi co lien thong khong? " 
         << (graph1.isConnected() ? "Co" : "Khong") << endl << endl;
    
    // Demo 2: Multigraph không liên thông
    cout << "2. MULTIGRAPH KHONG LIEN THONG:" << endl;
    auto graph2 = MultigraphUtils::createSampleMultigraph2();
    
    graph2.printGraph();
    graph2.printGraphStats();
    graph2.printMultipleEdges();
    graph2.bfsComplete();
    
    cout << "Do thi co lien thong khong? " 
         << (graph2.isConnected() ? "Co" : "Khong") << endl << endl;
    
    // Demo 3: Kiểm tra tính chất đặc biệt
    cout << "3. KIEM TRA TINH CHAT DAC BIET:" << endl;
    cout << "So luong canh giua dinh 0 va 1 trong graph1: " 
         << graph1.getEdgeMultiplicity(0, 1) << endl;
    cout << "So luong canh giua dinh 3 va 4 trong graph2: " 
         << graph2.getEdgeMultiplicity(3, 4) << endl;
}

int main() {
    demonstrateMultigraphBFS();
    
    // Demo tương tác
    cout << "=== DEMO TUONG TAC - MULTIGRAPH ===" << endl;
    int n, m;
    cout << "Nhap so dinh va so cap canh: ";
    cin >> n >> m;
    
    if (n <= 0 || m < 0) {
        cout << "So dinh va canh phai la so duong!" << endl;
        return 1;
    }
    
    MultigraphBFS userGraph(n);
    
    cout << "Nhap " << m << " cap canh (u v) - co the nhap trung de tao canh boi:" << endl;
    for (int i = 0; i < m; i++) {
        int u, v;
        cout << "Canh " << i + 1 << ": ";
        cin >> u >> v;
        
        if (u >= 0 && u < n && v >= 0 && v < n) {
            userGraph.addEdge(u, v);
        } else {
            cout << "Canh khong hop le! Dinh phai tu 0 den " << n-1 << endl;
            i--; // Nhập lại
        }
    }
    
    int start;
    cout << "Nhap dinh bat dau BFS: ";
    cin >> start;
    
    if (start < 0 || start >= n) {
        cout << "Dinh bat dau khong hop le!" << endl;
        return 1;
    }
    
    cout << "\n=== KET QUA ===" << endl;
    userGraph.printGraph();
    userGraph.printGraphStats();
    userGraph.printMultipleEdges();
    userGraph.bfs(start);
    userGraph.printDistances(start);
    userGraph.printBFSTree(start);
    
    return 0;
}

File: combinatorics/resource/LCHP/problem_LTDT_code/bai_9/bai9.py
from collections import deque, defaultdict
from typing import List, Dict, Set, Tuple, Optional

class MultigraphBFS:
    """BFS cho multigraph (đa đồ thị)"""
    
    def __init__(self, num_vertices: int):
        self.num_vertices = num_vertices
        # Sử dụng list để lưu nhiều cạnh giữa hai đỉnh
        self.adj_list = [[] for _ in range(num_vertices)]
        self.edge_count = defaultdict(int)  # Đếm số cạnh giữa hai đỉnh
        self.reset_arrays()
    
    def reset_arrays(self):
        """Reset các mảng trạng thái"""
        self.visited = [False] * self.num_vertices
        self.distance = [float('inf')] * self.num_vertices
        self.parent = [-1] * self.num_vertices
        self.bfs_order = []
    
    def add_edge(self, u: int, v: int):
        """Thêm cạnh vào multigraph"""
        if not (0 <= u < self.num_vertices and 0 <= v < self.num_vertices):
            print(f"Cạnh ({u},{v}) không hợp lệ!")
            return
            
        self.adj_list[u].append(v)
        self.adj_list[v].append(u)
        
        # Đếm số cạnh giữa u và v
        edge = tuple(sorted([u, v]))
        self.edge_count[edge] += 1
        
        print(f"Thêm cạnh ({u},{v}) - Số cạnh giữa {u} và {v}: {self.edge_count[edge]}")
    
    def bfs(self, start: int) -> List[int]:
        """
        Thực hiện BFS trên multigraph
        
        Args:
            start: Đỉnh bắt đầu
            
        Returns:
            Thứ tự duyệt BFS
        """
        if not (0 <= start < self.num_vertices):
            print(f"Đỉnh bắt đầu {start} không hợp lệ!")
            return []
            
        self.reset_arrays()
        
        queue = deque([start])
        self.visited[start] = True
        self.distance[start] = 0
        
        print(f"BFS trên Multigraph từ đỉnh {start}:")
        print("Thứ tự duyệt: ", end="")
        
        while queue:
            u = queue.popleft()
            self.bfs_order.append(u)
            print(u, end=" ")
            
            # Lấy danh sách đỉnh kề duy nhất để tránh duyệt trùng
            # Sử dụng set để loại bỏ trùng lặp
            unique_neighbors = set(self.adj_list[u])
            
            for v in sorted(unique_neighbors):
                if not self.visited[v]:
                    self.visited[v] = True
                    self.distance[v] = self.distance[u] + 1
                    self.parent[v] = u
                    queue.append(v)
        
        print("\n")
        return self.bfs_order.copy()
    
    def bfs_complete(self) -> Dict[int, List[int]]:
        """
        BFS toàn bộ đồ thị (xử lý thành phần liên thông)
        
        Returns:
            Dictionary chứa các thành phần liên thông
        """
        self.reset_arrays()
        components = {}
        component_count = 0
        
        print("BFS toàn bộ multigraph:")
        
        for i in range(self.num_vertices):
            if not self.visited[i]:
                component_count += 1
                component = self._bfs_component(i)
                components[component_count] = component
                print(f"Thành phần liên thông {component_count}: {' '.join(map(str, component))}")
        
        print(f"Tổng số thành phần liên thông: {component_count}\n")
        return components
    
    def _bfs_component(self, start: int) -> List[int]:
        """BFS cho một thành phần liên thông"""
        component = []
        queue = deque([start])
        self.visited[start] = True
        self.distance[start] = 0
        
        while queue:
            u = queue.popleft()
            component.append(u)
            
            unique_neighbors = set(self.adj_list[u])
            for v in sorted(unique_neighbors):
                if not self.visited[v]:
                    self.visited[v] = True
                    self.distance[v] = self.distance[u] + 1
                    self.parent[v] = u
                    queue.append(v)
        
        return component
    
    def get_shortest_path(self, start: int, target: int) -> List[int]:
        """
        Tìm đường đi ngắn nhất từ start đến target
        
        Args:
            start: Đỉnh bắt đầu
            target: Đỉnh đích
            
        Returns:
            Đường đi ngắn nhất (rỗng nếu không có đường đi)
        """
        self.bfs(start)
        
        if self.distance[target] == float('inf'):
            return []  # Không có đường đi
        
        # Tái tạo đường đi từ mảng parent
        path = []
        current = target
        while current != -1:
            path.append(current)
            current = self.parent[current]
        
        path.reverse()
        return path
    
    def print_distances(self, start: int):
        """In thông tin khoảng cách"""
        print(f"Khoảng cách từ đỉnh {start}:")
        print(f"{'Đỉnh':<6} {'Khoảng cách':<12} {'Cha':<10}")
        print("-" * 28)
        
        for i in range(self.num_vertices):
            dist_str = "INF" if self.distance[i] == float('inf') else str(self.distance[i])
            print(f"{i:<6} {dist_str:<12} {self.parent[i]:<10}")
        print()
    
    def print_bfs_tree(self, start: int):
        """In cây BFS"""
        print(f"Cây BFS từ đỉnh {start}:")
        
        tree_edges = []
        for i in range(self.num_vertices):
            if self.parent[i] != -1:
                tree_edges.append((self.parent[i], i))
        
        if tree_edges:
            for parent, child in tree_edges:
                print(f"{parent} -> {child}")
        else:
            print("Không có cây BFS (đồ thị rỗng hoặc chỉ có một đỉnh)")
        print()
    
    def print_multiple_edges(self):
        """In thông tin về các cạnh bội"""
        print("Các cạnh bội trong multigraph:")
        has_multiple = False
        
        for edge, count in sorted(self.edge_count.items()):
            if count > 1:
                print(f"Cạnh {edge}: {count} cạnh")
                has_multiple = True
        
        if not has_multiple:
            print("Không có cạnh bội")
        print()
    
    def get_edge_multiplicity(self, u: int, v: int) -> int:
        """Lấy số lượng cạnh giữa u và v"""
        edge = tuple(sorted([u, v]))
        return self.edge_count.get(edge, 0)
    
    def print_graph(self):
        """In biểu diễn multigraph"""
        print("Biểu diễn Multigraph (danh sách kề):")
        for i in range(self.num_vertices):
            neighbors = ' '.join(map(str, self.adj_list[i]))
            print(f"{i}: {neighbors}")
        print()
    
    def get_total_edges(self) -> int:
        """Tính tổng số cạnh (kể cả cạnh bội)"""
        return sum(self.edge_count.values())
    
    def is_connected(self) -> bool:
        """Kiểm tra đồ thị có liên thông không"""
        if self.num_vertices == 0:
            return True
            
        self.bfs(0)
        return all(self.visited)
    
    def find_all_paths_between(self, start: int, end: int, max_length: int = 10) -> List[List[int]]:
        """
        Tìm tất cả đường đi từ start đến end (có giới hạn độ dài)
        
        Args:
            start: Đỉnh bắt đầu
            end: Đỉnh kết thúc  
            max_length: Độ dài đường đi tối đa
            
        Returns:
            Danh sách các đường đi
        """
        all_paths = []
        
        def dfs_paths(current: int, target: int, path: List[int], visited: Set[int]):
            if len(path) > max_length:
                return
                
            if current == target:
                all_paths.append(path.copy())
                return
            
            # Lấy đỉnh kề duy nhất
            unique_neighbors = set(self.adj_list[current])
            for neighbor in unique_neighbors:
                if neighbor not in visited:
                    path.append(neighbor)
                    visited.add(neighbor)
                    dfs_paths(neighbor, target, path, visited)
                    path.pop()
                    visited.remove(neighbor)
        
        if start == end:
            return [[start]]
        
        dfs_paths(start, end, [start], {start})
        return all_paths
    
    def get_graph_statistics(self) -> Dict:
        """Lấy thống kê về đồ thị"""
        stats = {
            'num_vertices': self.num_vertices,
            'total_edges': self.get_total_edges(),
            'unique_edges': len(self.edge_count),
            'multiple_edges': sum(1 for count in self.edge_count.values() if count > 1),
            'is_connected': self.is_connected(),
            'components': len(self.bfs_complete())
        }
        return stats


class MultigraphUtils:
    """Các hàm tiện ích cho Multigraph"""
    
    @staticmethod
    def create_sample_multigraph1() -> MultigraphBFS:
        """Tạo multigraph mẫu 1"""
        graph = MultigraphBFS(5)
        
        # Thêm cạnh bình thường
        graph.add_edge(0, 1)
        graph.add_edge(1, 2)
        graph.add_edge(2, 3)
        graph.add_edge(3, 4)
        
        # Thêm cạnh bội
        graph.add_edge(0, 1)  # Cạnh thứ 2 giữa 0 và 1
        graph.add_edge(1, 2)  # Cạnh thứ 2 giữa 1 và 2
        graph.add_edge(0, 2)  # Cạnh mới
        
        return graph
    
    @staticmethod
    def create_sample_multigraph2() -> MultigraphBFS:
        """Tạo multigraph mẫu 2 - không liên thông"""
        graph = MultigraphBFS(7)
        
        # Thành phần 1
        graph.add_edge(0, 1)
        graph.add_edge(0, 1)  # Cạnh bội
        graph.add_edge(1, 2)
        
        # Thành phần 2
        graph.add_edge(3, 4)
        graph.add_edge(3, 4)  # Cạnh bội
        graph.add_edge(4, 5)
        graph.add_edge(3, 5)
        
        # Đỉnh cô lập
        # Đỉnh 6 không có cạnh nào
        
        return graph
    
    @staticmethod
    def create_complete_multigraph(n: int, multiplicity: int = 2) -> MultigraphBFS:
        """Tạo đồ thị đầy đủ với cạnh bội"""
        graph = MultigraphBFS(n)
        
        for i in range(n):
            for j in range(i + 1, n):
                for _ in range(multiplicity):
                    graph.add_edge(i, j)
        
        return graph


def demonstrate_multigraph_bfs():
    """Demo các tính năng của Multigraph BFS"""
    print("=== DEMO BFS TRÊN MULTIGRAPH ===\n")
    
    # Demo 1: Multigraph liên thông
    print("1. MULTIGRAPH LIÊN THÔNG:")
    graph1 = MultigraphUtils.create_sample_multigraph1()
    graph1.print_graph()
    graph1.print_multiple_edges()
    
    # BFS từ đỉnh 0
    graph1.bfs(0)
    graph1.print_distances(0)
    graph1.print_bfs_tree(0)
    
    # Tìm đường đi ngắn nhất
    print("Đường đi ngắn nhất từ 0 đến 4:")
    path = graph1.get_shortest_path(0, 4)
    if path:
        path_str = ' -> '.join(map(str, path))
        print(f"{path_str} (Độ dài: {len(path) - 1})")
    else:
        print("Không có đường đi")
    print()
    
    # Thống kê đồ thị
    stats1 = graph1.get_graph_statistics()
    print("Thống kê đồ thị 1:")
    for key, value in stats1.items():
        print(f"  {key}: {value}")
    print()
    
    # Demo 2: Multigraph không liên thông
    print("2. MULTIGRAPH KHÔNG LIÊN THÔNG:")
    graph2 = MultigraphUtils.create_sample_multigraph2()
    graph2.print_graph()
    graph2.print_multiple_edges()
    graph2.bfs_complete()
    
    stats2 = graph2.get_graph_statistics()
    print("Thống kê đồ thị 2:")
    for key, value in stats2.items():
        print(f"  {key}: {value}")
    print()
    
    # Demo 3: Complete multigraph
    print("3. COMPLETE MULTIGRAPH (4 đỉnh, mỗi cạnh có 2 bản sao):")
    complete_graph = MultigraphUtils.create_complete_multigraph(4, 2)
    complete_graph.print_graph()
    complete_graph.print_multiple_edges()
    complete_graph.bfs(0)
    
    stats3 = complete_graph.get_graph_statistics()
    print("Thống kê complete multigraph:")
    for key, value in stats3.items():
        print(f"  {key}: {value}")
    print()


def interactive_demo():
    """Demo tương tác với người dùng"""
    print("=== DEMO TƯƠNG TÁC MULTIGRAPH ===")
    
    try:
        n = int(input("Nhập số đỉnh: "))
        if n <= 0:
            print("Số đỉnh phải là số dương!")
            return
        
        m = int(input("Nhập số cạnh: "))
        if m < 0:
            print("Số cạnh không thể âm!")
            return
        
        user_graph = MultigraphBFS(n)
        
        print(f"Nhập {m} cạnh (u v), có thể có cạnh trùng lặp:")
        for i in range(m):
            try:
                u, v = map(int, input(f"Cạnh {i+1}: ").split())
                if 0 <= u < n and 0 <= v < n:
                    user_graph.add_edge(u, v)
                else:
                    print(f"Cạnh ({u}, {v}) không hợp lệ! Đỉnh phải từ 0 đến {n-1}")
                    return
            except ValueError:
                print("Vui lòng nhập hai số nguyên!")
                return
        
        start = int(input("Nhập đỉnh bắt đầu BFS: "))
        if not (0 <= start < n):
            print(f"Đỉnh bắt đầu phải từ 0 đến {n-1}!")
            return
        
        print("\nKết quả:")
        user_graph.print_graph()
        user_graph.print_multiple_edges()
        user_graph.bfs(start)
        user_graph.print_distances(start)
        user_graph.print_bfs_tree(start)
        
        # Thống kê
        stats = user_graph.get_graph_statistics()
        print("Thống kê đồ thị:")
        for key, value in stats.items():
            print(f"  {key}: {value}")
        
        # Tùy chọn bổ sung
        while True:
            print("\nCác tùy chọn bổ sung:")
            print("1. Tìm đường đi ngắn nhất")
            print("2. Hiển thị tất cả thành phần liên thông")
            print("3. Kiểm tra số lượng cạnh giữa hai đỉnh")
            print("4. Tìm tất cả đường đi giữa hai đỉnh")
            print("0. Thoát")
            
            choice = input("Chọn tùy chọn (0-4): ").strip()
            
            if choice == '0':
                break
            elif choice == '1':
                target = int(input("Nhập đỉnh đích: "))
                if 0 <= target < n:
                    path = user_graph.get_shortest_path(start, target)
                    if path:
                        path_str = ' -> '.join(map(str, path))
                        print(f"Đường đi ngắn nhất từ {start} đến {target}: {path_str}")
                        print(f"Độ dài: {len(path) - 1}")
                    else:
                        print(f"Không có đường đi từ {start} đến {target}")
                else:
                    print("Đỉnh đích không hợp lệ!")
            
            elif choice == '2':
                components = user_graph.bfs_complete()
                print(f"Số thành phần liên thông: {len(components)}")
            
            elif choice == '3':
                u = int(input("Nhập đỉnh u: "))
                v = int(input("Nhập đỉnh v: "))
                if 0 <= u < n and 0 <= v < n:
                    count = user_graph.get_edge_multiplicity(u, v)
                    print(f"Số cạnh giữa {u} và {v}: {count}")
                else:
                    print("Đỉnh không hợp lệ!")
            
            elif choice == '4':
                u = int(input("Nhập đỉnh bắt đầu: "))
                v = int(input("Nhập đỉnh kết thúc: "))
                if 0 <= u < n and 0 <= v < n:
                    paths = user_graph.find_all_paths_between(u, v)
                    if paths:
                        print(f"Tất cả đường đi từ {u} đến {v}:")
                        for i, path in enumerate(paths, 1):
                            path_str = ' -> '.join(map(str, path))
                            print(f"  {i}: {path_str}")
                    else:
                        print(f"Không có đường đi từ {u} đến {v}")
                else:
                    print("Đỉnh không hợp lệ!")
            
            else:
                print("Tùy chọn không hợp lệ!")
    
    except ValueError:
        print("Vui lòng nhập số nguyên hợp lệ!")
    except KeyboardInterrupt:
        print("\nChương trình bị ngắt bởi người dùng.")


if __name__ == "__main__":
    # Chạy demo tự động
    demonstrate_multigraph_bfs()
    
    # Chạy demo tương tác
    print("\n" + "="*60)
    interactive_demo()